lib/matplotlib/bezier.py

```diff
@@ -168,27 +176,113 @@ def find_bezier_t_intersecting_with_closedpath(
 
 class BezierSegment:
     """
-    A D-dimensional Bezier segment.
+    A d-dimensional Bezier segment.
 
     Parameters
     ----------
-    control_points : (N, D) array
+    control_points : (N, d) array
         Location of the *N* control points.
     """
 
     def __init__(self, control_points):
-        n = len(control_points)
-        self._orders = np.arange(n)
-        coeff = [math.factorial(n - 1)
-                 // (math.factorial(i) * math.factorial(n - 1 - i))
-                 for i in range(n)]
-        self._px = np.asarray(control_points).T * coeff
+        self._cpoints = np.asarray(control_points)
+        self._N, self._d = self._cpoints.shape
+        self._orders = np.arange(self._N)
+        coeff = [math.factorial(self._N - 1)
+                 // (math.factorial(i) * math.factorial(self._N - 1 - i))
+                 for i in range(self._N)]
+        self._px = self._cpoints.T * coeff
 
     def point_at_t(self, t):
         """Return the point on the Bezier curve for parameter *t*."""
         return tuple(
             self._px @ (((1 - t) ** self._orders)[::-1] * t ** self._orders))
 
+    @property
+    def control_points(self):
+        """The control points of the curve."""
+        return self._cpoints
+
+    @property
+    def dimension(self):
+        """The dimension of the curve."""
+        return self._d
+
+    @property
+    def degree(self):
+        """The number of control points in the curve."""
+        return self._N - 1
+
+    @property
+    def polynomial_coefficients(self):
+        r"""
+        The polynomial coefficients of the Bezier curve.
+
+        Returns
+        -------
+        float, (n+1, d) array_like
+            Coefficients after expanding in polynomial basis, where :math:`n`
+            is the degree of the bezier curve and :math:`d` its dimension.
+            These are the numbers (:math:`C_j`) such that the curve can be
+            written :math:`\sum_{j=0}^n C_j t^j`.
+
+        Notes
+        -----
+        The coefficients are calculated as
+
+        .. math::
+
+            {n \choose j} \sum_{i=0}^j (-1)^{i+j} {j \choose i} P_i
+
+        where :math:`P_i` are the control points of the curve.
+        """
+        n = self.degree
+        # matplotlib uses n <= 4. overflow plausible starting around n = 15.
+        if n > 10:
+            warnings.warn("Polynomial coefficients formula unstable for high "
+                          "order Bezier curves!", RuntimeWarning)
+        d = self.dimension
+        P = self.control_points
+        coefs = np.zeros((n+1, d))
+        for j in range(n+1):
+            i = np.arange(j+1)
+            prefactor = np.power(-1, i + j) * _comb(j, i)
+            coefs[j] = _comb(n, j) * np.sum(prefactor[:, None]*P[i], axis=0)
```

Comment:
Was going to mostly review the area PR and leave this one to the others but I guess the following comment applies throughout the stack: would it make sense to make _comb vectorized (even if naively by using `np.vectorize`) and later use vector ops where possible?

Reply:
But `_comb` is vectorized? `i` is an array, but since its length depends on which `j` you're using, I'm not sure you can vectorize over `j`.

Reply:
`_comb` is already vectorized, and is in fact called in a vectorized way to compute `prefactor` above.

I originally wrote it like this, with one for loop and everything else vectorized, for a couple of reasons
1) additional vectorization is unlikely to improve performance (since the calls to `np.power`, `_comb`, and `np.sum` are already vectorized). in fact, since you'd have to construct a 
2) my personal opinion is that the for loop is more readable. otherwise you have to keep track of some pretty complicated array broadcasting/reductions along different axes (like keeping a n^2 mask around to do the partial sum of i from 1 to j+1).

Reply:
My suggestion was to have _comb be "naively" vectorized (e.g. using np.vectorize, possibly stacked over lru_cache given that the number of reasonable arguments with which we call it is pretty limited to start with).  *If* it is the case *and* `comb(n, p)` is implemented to return 0 when p>n (the standard definition), then I would write the thing as
```python
j = np.arange(n+1)[:, None]
i = np.arange(n)[None, :]  # only i <= j will be nonzero
prefactor = (-1)**(i+j) * _comb(i, j)  # j along axis 0, i along axis 1
pP = prefactor @ P  # j along axis 0, d along axis 1
return _comb(n, j.ravel()) * pP
```
but am not insisting on it.
(At least the `power` call can be `(-1)**(i+j)`, I assume numpy implements `__pow__` (or rather `__rpow__`) in terms of `power`).

Reply:
Ah, the key thing I wasn't clicking on was defining `_comb(n, p)` to return zero for $p > n$. In that case all the masks I was worried about being confusing aren't a problem anymore. 

Since you went through the trouble of outlining it, I went ahead and tidied it up:

```
        j = np.arange(n+1)[:, None]
        i = np.arange(n+1)[None, :]
        prefactor = (-1)**(i + j) * _comb(j, i)
        return _comb(n, j) * prefactor @ P
```

Doing the vectorization is about a 3x speedup (includes LRU cache)
```
In [45]: curves = [matplotlib.bezier.BezierSegment(np.random.random_sample((4, 2))) for _ in range(1000)]
In [46]: %time x=[curve.polynomial_coefficients for curve in curves];
CPU times: user 172 ms, sys: 12 ms, total: 184 ms
Wall time: 178 ms

In [47]: %time x=[curve.vectorized_poly_coef for curve in curves];
CPU times: user 48 ms, sys: 4 ms, total: 52 ms
Wall time: 50.5 ms
```
 
and removing the LRU cache shows it is also around a 4x speedup:
```
In [53]: %time x=[curve.vectorized_no_cache for curve in curves];
CPU times: user 196 ms, sys: 4 ms, total: 200 ms
Wall time: 194 ms
```